### DISCUSSION TITLE
Some question about flush/merge.

### DISCUSSION THREAD

**Discussion by lajiyuan** [milvus-io/milvus#5301] [Q&A and General discussion]
https://github.com/milvus-io/milvus/blob/1.1/core/src/db/DBImpl.cpp#L2610

The note here says that all collections will be flushed, but the empty collection is used as a parameter. Why?

**Comment by shengjun1985** [comment]
It is due to WAL's interface design. Please refer:
https://github.com/milvus-io/milvus/blob/1.1/core/src/db/wal/WalManager.h#L152

**Reply by lajiyuan** [reply]
Thanks.
I am looking at the code for the background merge task. I add some log at https://github.com/milvus-io/milvus/blob/1.1/core/src/db/DBImpl.cpp#L1994.The log will be printed every second, but every time the collection is empty.

In fact, my collection information is ：
{'partitions': [{'row_count': 43571691, 'segments': [{'data_size': 317254895, 'index_name': 'IVFSQ8', 'name': '1621241748029890000', 'row_count': 593906}, {'data_size': 317254895, 'index_name': 'IVFSQ8', 'name': '1621241748029890000', 'row_count': 593906}, {'data_size': 285422047, 'index_name': 'IVFSQ8', 'name': '1621241801115374000', 'row_count': 532690}, {'data_size': 285422047, 'index_name': 'IVFSQ8', 'name': '1621241801115374000', 'row_count': 532690}, {'data_size': 335543463, 'index_name': 'IVFSQ8', 'name': '1621241853073160000', 'row_count': 629077}, {'data_size': 335543463, 'index_name': 'IVFSQ8', 'name': '1621241853073160000', 'row_count': 629077}, {'data_size': 280486559, 'index_name': 'IVFSQ8', 'name': '1621242008899244000', 'row_count': 523194}, {'data_size': 280486559, 'index_name': 'IVFSQ8', 'name': '1621242008899244000', 'row_count': 523194}, {'data_size': 323624279, 'index_name': 'IVFSQ8', 'name': '1621242055021703000', 'row_count': 606155}, {'data_size': 323624279, 'index_name': 'IVFSQ8', 'name': '1621242055021703000', 'row_count': 606155}, {'data_size': 280574311, 'index_name': 'IVFSQ8', 'name': '1621242101289915000', 'row_count': 523367}, {'data_size': 280574311, 'index_name': 'IVFSQ8', 'name': '1621242101289915000', 'row_count': 523367}, {'data_size': 311721807, 'index_name': 'IVFSQ8', 'name': '1621242150040970000', 'row_count': 583266}, {'data_size': 329678919, 'index_name': 'IVFSQ8', 'name': '1621242219966404000', 'row_count': 617799}, {'data_size': 311721807, 'index_name': 'IVFSQ8', 'name': '1621242150040970000', 'row_count': 583266}, {'data_size': 323080799, 'index_name': 'IVFSQ8', 'name': '1621242359646761000', 'row_count': 605106}, {'data_size': 329678919, 'index_name': 'IVFSQ8', 'name': '1621242219966404000', 'row_count': 617799}, {'data_size': 318039551, 'index_name': 'IVFSQ8', 'name': '1621242560202255000', 'row_count': 595474}, {'data_size': 63673159, 'index_name': 'IVFSQ8', 'name': '1621242594485224000', 'row_count': 111467}, {'data_size': 4817199, 'index_name': 'IVFSQ8', 'name': '1621242597447180000', 'row_count': 8428}, {'data_size': 5194175, 'index_name': 'IVFSQ8', 'name': '1621242599808594000', 'row_count': 9086}, {'data_size': 5935679, 'index_name': 'IVFSQ8', 'name': '1621242603937772000', 'row_count': 10386}, {'data_size': 5548239, 'index_name': 'IVFSQ8', 'name': '1621242607250689000', 'row_count': 9708}, {'data_size': 4241303, 'index_name': 'IVFSQ8', 'name': '1621242610322297000', 'row_count': 7419}, {'data_size': 3754263, 'index_name': 'IVFSQ8', 'name': '1621242612950301000', 'row_count': 6565}, {'data_size': 323080799, 'index_name': 'IVFSQ8', 'name': '1621242359646761000', 'row_count': 605106}, {'data_size': 318039551, 'index_name': 'IVFSQ8', 'name': '1621242560202255000', 'row_count': 595474}, {'data_size': 63673159, 'index_name': 'IVFSQ8', 'name': '1621242594485224000', 'row_count': 111467}, {'data_size': 323216671, 'index_name': 'IVFSQ8', 'name': '1621243394707794000', 'row_count': 605430}, {'data_size': 323216671, 'index_name': 'IVFSQ8', 'name': '1621243394707794000', 'row_count': 605430}, {'data_size': 82299999, 'index_name': 'IVFSQ8', 'name': '1621243460796414000', 'row_count': 144078}, {'data_size': 82299999, 'index_name': 'IVFSQ8', 'name': '1621243460796414000', 'row_count': 144078}, {'data_size': 284328111, 'index_name': 'IVFSQ8', 'name': '1621244641606031000', 'row_count': 530644}, {'data_size': 466034831, 'index_name': 'IVFSQ8', 'name': '1621244642413714000', 'row_count': 880080}, {'data_size': 466034831, 'index_name': 'IVFSQ8', 'name': '1621244642413714000', 'row_count': 880080}, {'data_size': 385112431, 'index_name': 'IVFSQ8', 'name': '1621244665210460000', 'row_count': 724460}, {'data_size': 284328111, 'index_name': 'IVFSQ8', 'name': '1621244641606031000', 'row_count': 530644}, {'data_size': 385112431, 'index_name': 'IVFSQ8', 'name': '1621244664563263000', 'row_count': 724460}, {'data_size': 385112431, 'index_name': 'IVFSQ8', 'name': '1621244664563263000', 'row_count': 724460}, {'data_size': 385112431, 'index_name': 'IVFSQ8', 'name': '1621244665210460000', 'row_count': 724460}, {'data_size': 293796791, 'index_name': 'IVFSQ8', 'name': '1621244738539054000', 'row_count': 548853}, {'data_size': 293796791, 'index_name': 'IVFSQ8', 'name': '1621244738588471000', 'row_count': 548853}, {'data_size': 293796791, 'index_name': 'IVFSQ8', 'name': '1621244738539054000', 'row_count': 548853}, {'data_size': 293796791, 'index_name': 'IVFSQ8', 'name': '1621244738588471000', 'row_count': 548853}, {'data_size': 283742071, 'index_name': 'IVFSQ8', 'name': '1621244743508284000', 'row_count': 529517}, {'data_size': 283742071, 'index_name': 'IVFSQ8', 'name': '1621244743508284000', 'row_count': 529517}, {'data_size': 390414351, 'index_name': 'IVFSQ8', 'name': '1621244771741798000', 'row_count': 734656}, {'data_size': 390414351, 'index_name': 'IVFSQ8', 'name': '1621244771741798000', 'row_count': 734656}, {'data_size': 295096271, 'index_name': 'IVFSQ8', 'name': '1621244772297132000', 'row_count': 551352}, {'data_size': 295096271, 'index_name': 'IVFSQ8', 'name': '1621244772297132000', 'row_count': 551352}, {'data_size': 317182751, 'index_name': 'IVFSQ8', 'name': '1621244778844683000', 'row_count': 593826}, {'data_size': 317182751, 'index_name': 'IVFSQ8', 'name': '1621244778844683000', 'row_count': 593826}, {'data_size': 303764671, 'index_name': 'IVFSQ8', 'name': '1621244803841825000', 'row_count': 568022}, {'data_size': 303764671, 'index_name': 'IVFSQ8', 'name': '1621244803841825000', 'row_count': 568022}, {'data_size': 303764671, 'index_name': 'IVFSQ8', 'name': '1621244805361546000', 'row_count': 568022}, {'data_size': 303764671, 'index_name': 'IVFSQ8', 'name': '1621244805361546000', 'row_count': 568022}, {'data_size': 348710351, 'index_name': 'IVFSQ8', 'name': '1621244820534095000', 'row_count': 654456}, {'data_size': 348710351, 'index_name': 'IVFSQ8', 'name': '1621244820534095000', 'row_count': 654456}, {'data_size': 348710351, 'index_name': 'IVFSQ8', 'name': '1621244821209376000', 'row_count': 654456}, {'data_size': 348710351, 'index_name': 'IVFSQ8', 'name': '1621244821209376000', 'row_count': 654456}, {'data_size': 397053711, 'index_name': 'IVFSQ8', 'name': '1621244842177321000', 'row_count': 747424}, {'data_size': 397053711, 'index_name': 'IVFSQ8', 'name': '1621244842177321000', 'row_count': 747424}, {'data_size': 397053711, 'index_name': 'IVFSQ8', 'name': '1621244842310842000', 'row_count': 747424}, {'data_size': 397053711, 'index_name': 'IVFSQ8', 'name': '1621244842310842000', 'row_count': 747424}, {'data_size': 357323631, 'index_name': 'IVFSQ8', 'name': '1621244874131922000', 'row_count': 671020}, {'data_size': 357323631, 'index_name': 'IVFSQ8', 'name': '1621244874131922000', 'row_count': 671020}, {'data_size': 357323631, 'index_name': 'IVFSQ8', 'name': '1621244875732512000', 'row_count': 671020}, {'data_size': 357323631, 'index_name': 'IVFSQ8', 'name': '1621244875732512000', 'row_count': 671020}, {'data_size': 381826031, 'index_name': 'IVFSQ8', 'name': '1621244907727312000', 'row_count': 718140}, {'data_size': 381826031, 'index_name': 'IVFSQ8', 'name': '1621244907727312000', 'row_count': 718140}, {'data_size': 381826031, 'index_name': 'IVFSQ8', 'name': '1621244908450369000', 'row_count': 718140}, {'data_size': 381826031, 'index_name': 'IVFSQ8', 'name': '1621244908450369000', 'row_count': 718140}, {'data_size': 386824271, 'index_name': 'IVFSQ8', 'name': '1621244936482889000', 'row_count': 727752}, {'data_size': 386824271, 'index_name': 'IVFSQ8', 'name': '1621244936482889000', 'row_count': 727752}, {'data_size': 386824271, 'index_name': 'IVFSQ8', 'name': '1621244936663340000', 'row_count': 727752}, {'data_size': 386824271, 'index_name': 'IVFSQ8', 'name': '1621244936663340000', 'row_count': 727752}, {'data_size': 284384271, 'index_name': 'IVFSQ8', 'name': '1621244962089303000', 'row_count': 530752}, {'data_size': 8127983, 'index_name': 'IVFSQ8', 'name': '1621244969765351000', 'row_count': 14224}, {'data_size': 284384271, 'index_name': 'IVFSQ8', 'name': '1621244962089303000', 'row_count': 530752}, {'data_size': 8127983, 'index_name': 'IVFSQ8', 'name': '1621244969765351000', 'row_count': 14224}, {'data_size': 7023296, 'index_name': 'IVFSQ8', 'name': '1621410921890208000', 'row_count': 3416}, {'data_size': 3024376, 'index_name': 'IVFSQ8', 'name': '1621410921891345000', 'row_count': 1471}, {'data_size': 5312704, 'index_name': 'IVFSQ8', 'name': '1621410922911743000', 'row_count': 2584}], 'tag': '_default'}], 'row_count': 43571691}

Can you give me some guidance as to why this is happening? 

In my understanding, MergeLayeredStrategy merges files of different layers every flush. But I still see that some small files are not merged when the system does not have any inserts and queries. Like this:
![image](https://user-images.githubusercontent.com/38530078/118791499-580b3980-b8c9-11eb-9552-8b67c5970de7.png)


Hope to get your reply, thank you！

**Comment by yhmo** [comment]
About merge strategy, please refer this article: https://zhuanlan.zhihu.com/p/148106312

"Layered" merge firstly group the un-merged files according to their size, each group is a layer. Then merge files within layer.
If no more insert action come in, there will be some small files not merged. These small files will be force-merge until the client call create_index() or the server restart.

**Reply by lajiyuan** [reply]
这篇文章我之前有读过，根据文章里所说的，应该每次落盘任务都会做一次层级合并策略，那么上述很多的20M- 30M的数据段应该被合并到一起。
但实际上我在层级合并策略内部加日志发现并没有文件合并的逻辑，磁盘上的20M~30M的数据端也没有被合并， 想问下这是为什么？

**Reply by yhmo** [reply]
你的insert模式是怎样的？比如每次插入多少数据，每分钟调用多少次insert

**Reply by lajiyuan** [reply]
在我观察合并策略的时候，我关掉了所有的client，理论上没有数据在插入。 

在之前测试的时候，插入速度大概为5000/s.

**Reply by yhmo** [reply]
我测了一下，index_file_size=2GB，512维向量，每次插入1000条，一共插入200万条向量，也就是4GB的数据，观察硬盘上下的数据是这样的：
```
2.2G	./1621423411085959000
1.2G	./1621423508595745000
299M	./1621423532370554000
156M	./1621423545436700000
26M	./1621423545393583000
2.0M	./1621423545471629000
3.9G	.
```
没有这么多小文件，是不是你的测试方法不太一样？仅仅只调用了insert()么？

**Reply by lajiyuan** [reply]
可能和我之前起了两个共享db和存储的milvus服务，并且没有用集群模式配置只读节点有关系？ 我再重新配置成一读一写测试一下看看。如果有问题再来找你。

**Reply by lajiyuan** [reply]
感谢回复

**Reply by yhmo** [reply]
如果多个milvus服务使用同一个db目录和同一个元数据服务，那么就只能有一个写节点，否则会出错。

**Comment by yhmo** [comment]
I don't understand "every time the collection is empty", could you explain more?

**Reply by lajiyuan** [reply]
> I don't understand "every time the collection is empty", could you explain more?
https://github.com/milvus-io/milvus/blob/1.1/core/src/db/DBImpl.cpp#L1994 

我在这里增加日志打印出merge_collection_ids，但是发现每秒的日志中这个变量都是空的，结合上面说的给我的感觉是合并任务一直没有被调用，是因为还有其他的触发条件吗？

**Reply by yhmo** [reply]
StartMergeTask() 会被多个地方调到，在flush的时候，server start的时候，drop/create index的时候会被触发。后台auto flush的线程会定时触发它，但是因为auto flush时不一定真的有数据落盘，所以merge_collection_ids是空的。当flush真的有数据落盘时，这个参数就不为空。

**Reply by lajiyuan** [reply]
非常感谢，我用正确的部署方式部署之后，每个数据段基本都是不同层下的，看上去符合预期了。 谢谢。
